File: java/bypstest/src/com/wilutions/byps/test/TestRemotePrimitiveTypes.java
package com.wilutions.byps.test;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import com.wilutions.byps.BException;
import com.wilutions.byps.test.api.BClient_Testser;
import com.wilutions.byps.test.api.prim.PrimitiveTypes;
import com.wilutions.byps.test.api.remote.RemotePrimitiveTypes;

/**
 * Tests for interface functions with primitive types.
 * This test requires the webapp bypstest-srv to be started. 
 * 
 *
 */
public class TestRemotePrimitiveTypes {
	
	BClient_Testser client;
	RemotePrimitiveTypes remote;
	private Log log = LogFactory.getLog(TestRemotePrimitiveTypes.class);

	@Before
	public void setUp() throws BException, InterruptedException {
		client = TestUtilsHttp.createClient();
		remote = client.remotePrimitiveTypes;
	}
	
	@After
	public void tearDown() {
		if (client != null) {
			client.done();
		}
	}
	
	@Test
	public void testRemotePrimitiveTypes() throws BException, InterruptedException {
		log.info("testRemotePrimitiveTypes(");
		
		remote.setBool(true);
		TestUtils.assertEquals(log, "bool", true, remote.getBool());
		remote.setBool(false);
		TestUtils.assertEquals(log, "bool", false, remote.getBool());
		remote.setByte((byte)11);
		TestUtils.assertEquals(log, "byte", (byte)11, remote.getByte());
		remote.setChar('€');
		TestUtils.assertEquals(log, "char", '€', remote.getChar());
		remote.setShort((short)1234);
		TestUtils.assertEquals(log, "short", (short)1234, remote.getShort());
		remote.setInt(5678);
		TestUtils.assertEquals(log,  "int", 5678, remote.getInt());
		remote.setLong(890L);
		TestUtils.assertEquals(log, "long", 890L, remote.getLong());
		remote.setFloat(1.2f);
		TestUtils.assertEquals(log,  "float", 1.2f, remote.getFloat());
		remote.setDouble(1.8);
		TestUtils.assertEquals(log,  "double", 1.8, remote.getDouble());
		remote.setString("ABC");
		TestUtils.assertEquals(log,  "String", "ABC", remote.getString());
		
		PrimitiveTypes pt = TestUtils.createObjectPrimitiveTypes();
		remote.setObject(pt);
		TestUtils.assertEquals(log,  "Object", pt, remote.getObject());
		
		pt = TestUtils.createObjectPrimitiveTypes();
		remote.setPrimitiveTypes(pt);
		TestUtils.assertEquals(log,  "PrimitiveTypes", pt, remote.getPrimitiveTypes());
		
		log.info(")testRemotePrimitiveTypes");
	}
	
	@Test
	public void testRemotePrimitiveTypesSendAll() throws BException, InterruptedException {
		log.info("testRemotePrimitiveTypesSendAll(");
		
		PrimitiveTypes pt = TestUtils.createObjectPrimitiveTypes();
		PrimitiveTypes pt2  =TestUtils.createObjectPrimitiveTypes();
		remote.sendAllTypes(true, 'Q', (short)34, 56, 45, 7.4f, 9.8, "23", pt, pt2);
		
		TestUtils.assertEquals(log, "bool", true, remote.getBool());
		TestUtils.assertEquals(log, "char", 'Q', remote.getChar());
		TestUtils.assertEquals(log, "short", (short)34, remote.getShort());
		TestUtils.assertEquals(log,  "int", 56, remote.getInt());
		TestUtils.assertEquals(log, "long", 45L, remote.getLong());
		TestUtils.assertEquals(log,  "float", 7.4f, remote.getFloat());
		TestUtils.assertEquals(log,  "double", 9.8, remote.getDouble());
		TestUtils.assertEquals(log,  "String", "23", remote.getString());
		TestUtils.assertEquals(log,  "PrimtitiveTypes", pt, remote.getPrimitiveTypes());
		TestUtils.assertEquals(log,  "Object", pt2, remote.getObject());

		log.info(")testRemotePrimitiveTypesSendAll");
	}
	
	@Test
	public void testPrimitiveTypesReferenceToOtherObject() throws BException, InterruptedException {	
		log.info("testPrimitiveTypesReferenceToOtherObject(");

		PrimitiveTypes obj1 = new PrimitiveTypes();
		obj1.intVal = 123;
		obj1.objVal = obj1.objVal2 = new PrimitiveTypes();
		((PrimitiveTypes)obj1.objVal).intVal = 456;
		
		remote.setPrimitiveTypes(obj1);
		PrimitiveTypes objR = remote.getPrimitiveTypes();
		
		TestUtils.assertEquals(log,  "PrimitiveTypes", obj1, objR);
		TestUtils.assertTrue(log, "this.objVal != this.objVal2", objR.objVal == objR.objVal2);
		
		log.info(")testPrimitiveTypesReferenceToOtherObject");
	}
	
	@Test
	public void testPrimitiveTypesReferenceToSelf() throws BException, InterruptedException {	
		log.info("testPrimitiveTypesReferenceToSelf(");

		PrimitiveTypes obj1 = new PrimitiveTypes();
		obj1.intVal = 123;
		obj1.objVal = obj1.objVal2 = obj1;
		
		remote.setPrimitiveTypes(obj1);
		PrimitiveTypes objR = remote.getPrimitiveTypes();
		
		TestUtils.assertTrue(log, "this != this.objVal", objR == objR.objVal);
		TestUtils.assertTrue(log, "this != this.objVal2", objR == objR.objVal2);
		
		log.info(")testPrimitiveTypesReferenceToSelf");
	}


}

File: csharp/bypstest/TestRemotePrimitiveTypes.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using byps;
using byps.test.api;
using byps.test.api.remote;
using byps.test.api.prim;
using byps.test.api.srvr;
using System.IO;
using System.Threading;
using System.Collections.Generic;

namespace bypstest
{
    [TestClass]
    public class TestRemotePrimitiveTypes
    {
        Log log = LogFactory.getLog(typeof(TestRemotePrimitiveTypes));
        BClient_Testser client;
        RemotePrimitiveTypes remote;

        [TestInitialize]
        public void setUp()
        {
            client = TestUtilsHttp.createClient();
            remote = client.RemotePrimitiveTypes;
        }

        [TestCleanup]
        public void tearDown()
        {
            client.done();
        }

        [TestMethod]
        public void TestPrimitiveTypes()
        {
            log.info("TestPrimitiveTypes(");
            try
            {
                remote.SetBool(true);
                Assert.AreEqual(true, remote.GetBool());
                remote.SetBool(false);
                TestUtils.assertEquals(log, "bool", false, remote.GetBool());
                remote.SetByte((byte)11);
                TestUtils.assertEquals(log, "byte", (byte)11, remote.GetByte());
                remote.SetChar('€');
                TestUtils.assertEquals(log, "char", '€', remote.GetChar());
                remote.SetShort((short)1234);
                TestUtils.assertEquals(log, "short", (short)1234, remote.GetShort());
                remote.SetInt(5678);
                TestUtils.assertEquals(log, "int", 5678, remote.GetInt());
                remote.SetLong(890L);
                TestUtils.assertEquals(log, "long", 890L, remote.GetLong());
                remote.SetFloat(1.2f);
                TestUtils.assertEquals(log, "float", 1.2f, remote.GetFloat());
                remote.SetDouble(1.8);
                TestUtils.assertEquals(log, "double", 1.8, remote.GetDouble());

                remote.SetString("ABC");
                String str = remote.GetString();
                TestUtils.assertEquals(log, "String", "ABC", str);

                PrimitiveTypes pt = TestUtils.createObjectPrimitiveTypes();
                remote.SetObject(pt);
                TestUtils.assertEquals(log, "Object", pt, remote.GetObject());

                pt = TestUtils.createObjectPrimitiveTypes();
                remote.SetPrimitiveTypes(pt);
                TestUtils.assertEquals(log, "PrimitiveTypes", pt, remote.GetPrimitiveTypes());
            }
            catch (BException e)
            {
                Assert.Fail(e.ToString());
            }

            log.info(")TestPrimitiveTypes");
        }

        [TestMethod]
        public void testRemotePrimitiveTypesSendAll()
        {
            log.info("testRemotePrimitiveTypesSendAll(");

            PrimitiveTypes pt = TestUtils.createObjectPrimitiveTypes();
            PrimitiveTypes pt2 = TestUtils.createObjectPrimitiveTypes();
            remote.SendAllTypes(true, 'Q', (short)34, 56, 45, 7.4f, 9.8, "23", pt, pt2);

            TestUtils.assertEquals(log, "bool", true, remote.GetBool());
            TestUtils.assertEquals(log, "char", 'Q', remote.GetChar());
            TestUtils.assertEquals(log, "short", (short)34, remote.GetShort());
            TestUtils.assertEquals(log, "int", 56, remote.GetInt());
            TestUtils.assertEquals(log, "long", 45L, remote.GetLong());
            TestUtils.assertEquals(log, "float", 7.4f, remote.GetFloat());
            TestUtils.assertEquals(log, "double", 9.8, remote.GetDouble());
            TestUtils.assertEquals(log, "String", "23", remote.GetString());
            TestUtils.assertEquals(log, "Object", pt2, remote.GetObject());
            TestUtils.assertEquals(log, "PrimtitiveTypes", pt, remote.GetPrimitiveTypes());

            log.info(")testRemotePrimitiveTypesSendAll");
        }

        [TestMethod]
        public void testRemotePrimitiveTypesAsync()
        {
            log.info("testRemotePrimitiveTypesAsync(");

            BSyncResult<Object> result = new BSyncResult<Object>();
            remote.SetInt(1122, (ignored, ex) =>
                {
                    if (ex != null) TestUtils.fail(log, ex.ToString());
                    remote.GetInt((valueR, ex2) =>
                        {
                            if (ex != null) TestUtils.fail(log, ex2.ToString());
                            TestUtils.assertEquals(log, "GetInt", 1122, valueR);
                        });
                });

            log.info(")testRemotePrimitiveTypesAsync");
        }

        /**
         * Tests asynchronous invocation with null as asyncResult parameter.
         * A null can be passed, if the result is not of interest.
         * @throws InterruptedException
         */
        [TestMethod]
        public void testRemotePrimitiveTypesAsyncNull()
        {
            log.info("testRemotePrimitiveTypesAsyncNull(");

            remote.SetInt(1122, null);
            Thread.Sleep(1000);

            log.info(")testRemotePrimitiveTypesAsyncNull");
        }

        [TestMethod]
        public void testPrimitiveTypesInt()
        {
            log.info("testPrimitiveTypesInt(");

            int[] arr = new int[] { 0, 1, 0xFF, 0x100, 0x10000, 0x1000000, 0x7FFFFFFF, -1, -2147483648 };
            for (int i = 1; i < arr.Length; i++)
            {
                remote.SetInt(arr[i]);
                int valueR = remote.GetInt();
                TestUtils.assertEquals(log, "Wrong int", arr[i], valueR);
            }
            log.info(")testPrimitiveTypesInt");
        }

        [TestMethod]
        public void testSerializeLong() 
        {
            log.info("testSerializeLong(");
            List<long> ints = new List<long>();
    
            ints.Add((long) 0);
            ints.Add((long) 1);
            ints.Add((long) 2);
 
            ulong a = 0;
            for (int i = 0; i < 63-7; i+=7) 
            {
                ints.Add((long)(((ulong)0x7E << i) | a));
                ints.Add((long)(((ulong)0x7F << i) | a));
                ints.Add((long)(((ulong)0x80 << i) | a));
                ints.Add((long)(((ulong)0x81 << i) | a));
              a <<= 7;
              a |= 0x5D;
            }

            for (int i = 0; i < ints.Count; i++) {
              internalTestLong(ints[i]);
              internalTestLong(-ints[i]);
            }
    
            internalTestLong(long.MaxValue);
            internalTestLong(long.MinValue);

            log.info(")testSerializeLong");
        }
  
        private void internalTestLong(long v) 
        {
            // Serialize/deserialize here
            BBufferBin bbuf = (BBufferBin)BBuffer.create(BBinaryModel.MEDIUM, null);
            bbuf.putLong(v);
            byte[] data = bbuf.getBuffer().array();
            BBufferBin bbuf2 = (BBufferBin)BBuffer.create(BBinaryModel.MEDIUM, ByteBuffer.wrap(data, 0, bbuf.position()));
            long r = bbuf2.getLong();
            TestUtils.assertEquals(log, "longVal", v, r);

            // Send to server, receive from server
            remote.SetLong(v);
            r = remote.GetLong();
            TestUtils.assertEquals(log, "Wrong long", v, r);
        }

    }

}
